cdap-master/src/main/java/io/cdap/cdap/master/upgrade/PostUpgradeJobMain.java

```diff
@@ -0,0 +1,104 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.cdap.cdap.master.upgrade;
+
+import io.cdap.cdap.client.ApplicationClient;
+import io.cdap.cdap.client.NamespaceClient;
+import io.cdap.cdap.client.ProgramClient;
+import io.cdap.cdap.client.ScheduleClient;
+import io.cdap.cdap.client.config.ClientConfig;
+import io.cdap.cdap.client.config.ConnectionConfig;
+import io.cdap.cdap.common.NotFoundException;
+import io.cdap.cdap.common.service.Retries;
+import io.cdap.cdap.common.service.RetryStrategies;
+import io.cdap.cdap.common.service.RetryStrategy;
+import io.cdap.cdap.proto.ApplicationRecord;
+import io.cdap.cdap.proto.NamespaceMeta;
+import io.cdap.cdap.proto.id.ApplicationId;
+import io.cdap.cdap.proto.id.NamespaceId;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * Restarts all schedules and programs stopped between startTimeMillis and now.
+ * The first parameter to this job should be the hostname, the second should be the port the
+ * router service is running on, and the third should be startTimeMillis.
+ * Eg, if the router is running at URI http://your-hostname:11015 and all schedules and programs stopped between
+ * 32432423324ms and now, then pass in
+ * your-hostname as the first parameter, 11015 as the second, and 32432423324 as the third.
+ */
+public class PostUpgradeJobMain {
+
+  private static final int DEFAULT_READ_TIMEOUT_MILLIS = 90 * 1000;
+
+  public static void main(String[] args) {
+    if (args.length != 3) {
+      throw new RuntimeException(
+        String.format("Invalid number of arguments to PostUpgradeJobMain. Needed 3, found %d", args.length));
+    }
+    ConnectionConfig connectionConfig = ConnectionConfig.builder()
+      .setHostname(args[0])
+      .setPort(Integer.parseInt(args[1]))
+      .setSSLEnabled(false)
+      .build();
+    ClientConfig clientConfig =
+      ClientConfig.builder()
+        .setDefaultReadTimeout(DEFAULT_READ_TIMEOUT_MILLIS)
+        .setConnectionConfig(connectionConfig)
+        .build();
+    RetryStrategy retryStrategy =
+      RetryStrategies.timeLimit(30, TimeUnit.SECONDS,
+                                RetryStrategies.exponentialDelay(10, 500, TimeUnit.MILLISECONDS));
+
+    try {
+      Retries.callWithRetries(
+        () -> {
+          restartPipelinesAndSchedules(clientConfig, Integer.parseInt(args[2]));
+          return null;
+        }, retryStrategy, e -> e instanceof IOException || e instanceof NotFoundException);
+    } catch (Exception e) {
+      throw new RuntimeException("Failed to restart pipelines and schedules.", e);
+    }
+  }
+
+  private static void restartPipelinesAndSchedules(ClientConfig clientConfig, long startTimeMillis) throws Exception {
```

Comment:
it's better to declare the exact exceptions that can be thrown, unless there's actually a method that throws Exception. Without this it's impossible to tell what type exceptions are not being retried above.

Reply:
namespaceClient.list throws Exception unfortunately